### ISSUE TITLE
PubSub example fails with logging enabled.

### ISSUE THREAD

**Issue by vaneek** [aws/aws-iot-device-sdk-arduino-yun#9]
Hi, I've tried to enable logging for the PubSub example.

I replaced the following code in the runtimeHub.py **init** function 
from;

```
    self._logManagerHub.disable()                                                                     
```

to;        

```
     self._logManagerHub.enable()                                                                     
     self._logManagerHub.enableFileOutput()  
```

The previously running example now fails consistently with a -1 error.

```
    AWS IoT SDK Version(dev) 1.0.3-

    Setup failed!
    -1
```

The logger does create a new log;

> ```
> [2016-03-04 14:10:41.256372] Register timeout signal handler.
> [2016-03-04 14:10:41.272485] serialCommunicationServer init.
> [2016-03-04 14:10:41.277257] serialCommunicationServer set accept timeout to 15
> [2016-03-04 14:10:41.282366] serialCommunicationServer set chunk size to 50
> [2016-03-04 14:10:41.287254] Clear internal list. Size: 0
> [2016-03-04 14:10:41.291974] Accept-timer starts, with acceptTimeout: 15 second(s).
> [2016-03-04 14:10:41.296875] 4 lines to be received. Loop begins.
> [2016-03-04 14:10:41.301735] Received: 1/4 Message is: i
> [2016-03-04 14:10:41.306648] Received: 2/4 Message is: cvmon1
> [2016-03-04 14:10:41.311479] Received: 3/4 Message is: 1
> [2016-03-04 14:10:41.316364] Received: 4/4 Message is: 4
> [2016-03-04 14:10:41.321048] Finish reading from remote client. Accept-timer ends.
> [2016-03-04 14:10:41.330683] Paho MQTT Client init.
> [2016-03-04 14:10:41.337009] Register Paho MQTT Client callbacks.
> [2016-03-04 14:10:41.341630] mqttCore init.
> [2016-03-04 14:10:41.346389] Set maximum connect/disconnect timeout to be 10 second.
> [2016-03-04 14:10:41.351113] Set maximum MQTT operation timeout to be 5 second
> [2016-03-04 14:10:41.356251] Updated serialCommunicationServer internal protocolMessageQueue by inserting a new message. Size: 1
> [2016-03-04 14:10:41.361421] Send through serial to remote client: I T Size: 3
> [2016-03-04 14:10:41.366106] Clear internal list. Size: 0
> [2016-03-04 14:10:41.370721] Accept-timer starts, with acceptTimeout: 15 second(s).
> [2016-03-04 14:10:56.370924] Raise a custom exception for accept timeout.
> [2016-03-04 14:10:56.373646] Accept Timeout
> ```

Have I missed something in enabling logging?

Thanks,

Jeff

**Comment by liuszeng** [comment]
Hi Jeff,

Thank you very much for using AWS IoT Arduino Yun SDK!

The AWS IoT Arduino Yun SDK is built on top of the dual processor architecture. ATmega32u4 accepts user commands and communicates them back to AR9331 through Serial1 bridge. The AR9331 then does all the heavy-lifting MQTT operations. A Serial1 communication protocol is established to make the transmission happen between the two processors, just as what you see in the log file.

This protocol occupies the STD_IN/STD_OUT on AR9331. In other words, the STD_IN/STD_OUT for AR9331 should not have other messages going through, otherwise it will break the protocol and trigger the error handling in SDK. The enable method for logging automatically enables the console output as default, which breaks the protocol. You will need to specifically disable it:

```
 self._logManagerHub.enable()
 self._logManagerHub.enableFileOutput()
 self._logManagerHub.disableConsolePrint()
```

In this case, you should be able to do persistent logging and store them as log files on AR9331, without breaking the Serial1 protocol in between.

However, one important thing to keep in mind is that doing persistent logging directly on AR9331 is not a very good idea, as is mentioned in the official [Arduino Yun Getting Started Guide](https://www.arduino.cc/en/Guide/ArduinoYun#toc4):

> You're discouraged from using the Yún's built-in non-volatile memory, because it has a limited number of writes.

If you really want persistent logging, one thing you can do is to use external memory like a microSD card or thumb drive and expand the disk space for Arduino Yun. More details can be found on [Arduino Tutorials](https://www.arduino.cc/en/Tutorial/ExpandingYunDiskSpace).

Hope that answers your question. Please feel free to reopen the issue if you have more questions.

Thanks,
Liusu

**Comment by vaneek** [comment]
Hi Liusu,

Thanks for the reply. I was aware of the Serial1 bridge but did not catch explicitly disabling the console. I have added a microSD card though I am only enabling the logging to extend the SDK to add some functionality lost by the unavailability of the Serial1 bridge. Specifically I would like to add a call to the python time() function in the runtime and return it to the mcu.

Is there any documentation available regarding extending the SDK?  

Thanks,
Jeff

**Comment by liuszeng** [comment]
Hi Jeff,

Currently we do not have a detailed documentation to extend the SDK.

For a custom command over the Serial1 bridge, the part of the SDK you need to extend is the [Python runtime library](https://github.com/aws/aws-iot-device-sdk-arduino-yun/tree/master/AWS-IoT-Python-Runtime).

The core functionality for AWS IoT Arduino Yun lies in the Python runtime library, which provides an interface for Serial1 communication: serialCommunicationServer. The AWS-IoT-Arduino-Yun-Library (the one gets included in Arduino IDE and runs on ATmega) generates the protocol-style command pieces and sends it over the Serial1 bridge. The serialCommunicationServer accepts commands (in protocol style) from ATmega and then triggers the corresponding command on the Python side.

-----------------------ATmega----------------------|-Serial1-|---------------------AR9331---------------------
ArduinoLib API call -> protocol-style command ->>>> serialCommunicationServer -> PyCmd -> AWS IoT
------------------------------------------------------|-----------|--------------------------------------------------

Each of the Serial1 command has its own protocol representation. You can get a better overview of the built-in commands by checking [here](https://github.com/aws/aws-iot-device-sdk-arduino-yun/tree/master/AWS-IoT-Python-Runtime/lib/command).

Take `aws_iot_mqtt_client::config` as an example. If you look into the implementation of [AWS-IoT-Arduino-Yun-Library](https://github.com/aws/aws-iot-device-sdk-arduino-yun/blob/master/AWS-IoT-Arduino-Yun-Library/aws_iot_mqtt.cpp), you can find that what `config` API actually does is to print protocol-style command pieces to Serial1:

```
exec_cmd("6\n", false, false);
exec_cmd("g\n", false, false);
    ...
exec_cmd(rw_buf, false, false);
exec_cmd(rw_buf, false, false);
    ...
exec_cmd(rw_buf, false, false);
    ...
exec_cmd(rw_buf, false, false);
    ...
exec_cmd(rw_buf, true, false);
```

So the protocol-style command pieces for `config` would like this:

```
6\ng\n<param0>\n<param1>\n<param2>\n<param3>\n<param4>\n
```

with a newline at the end of each chunk.

Then on the Python side (serialCommunicationServer), it accepts these protocol command pieces line by line and recognizes them to be the config command. It triggers the execution of the corresponding commandClass: [commandConfig](https://github.com/aws/aws-iot-device-sdk-arduino-yun/blob/master/AWS-IoT-Python-Runtime/lib/command/commandConfig.py), performs the operation and writes back the response:

```
...
self._mqttCoreHandler.config(self._parameterList[0], int(self._parameterList[1]), self._parameterList[2], self._parameterList[3], self._parameterList[4])
...
self._serialCommServerHandler.writeToInternalProtocol(returnMessage)
```

Basically, a custom command will do the same thing by utilizing the Linux processor and communicate back the resources to ATmega. Adding a new custom command requires a new Python command class, a new protocol command and the corresponding API for AWS-IoT-Arduino-Yun-Library. The serialCommunicationServer also needs to be modified to allow new commands to be accepted and executed.

We are sorry we cannot provide a more positive answer about extending the SDK at this moment. Please feel free to post follow-up questions as you dive deeper into this.

Thanks,
Liusu
